### PR TITLE
Auto-generate JSON Schema from TypeScript

### PR DESCRIPTION
**Pull request by alexiswl**
## Summary

1. Run [typescript-json-schema](https://www.npmjs.com/package/typescript-json-schema) over the Workflow schema.  
2. Refine the schema (see [.github/scripts/curate_autogenerated_schema.py](https://github.com/alexiswl/cwl-ts-auto/blob/enhancement/generate-json-schema/.github/scripts/curate_autogenerated_schema.py))

## Schema refining process steps

1. Remove loadingOptions and extensionFields properties

2. Rename `class_`, `type_` etc to `class`, `type`

3. Add `$import` and `$include` options and update requirements that allow for them (SchemaDefRequirementa and InlineJavascriptRequirement)

4. Fix inputs / outputs / steps, so not just an array but can be an object where the key is the `id` attribute.

5. Fix Requirements / Hints, so not just an array but can be an object where the key is the `class` attribute.

6. Fix descriptions by removing *Auto-generated...* from the start

7. Generate the CWLFile definition, which can be one of Workflow, CommandLineTool, or ExpressionTool and also fits the CWLMetadata

8. Fix issues with additionalProperties and nested definitions as shown here https://stoic-agnesi-d0ac4a.netlify.app/1

9. Write out the schema

10. Generate the CWLGraph definition, which is an array of CWLFile objects with CWLMetadata

11. Write out the graph schema to a separate file

### CODE REVIEW

**Comment by alexiswl** [comment]
## Installation

### JetBrains

Settings -> Languages & Frameworks -> Schemas and DTDs -> JSON Schema Mappings: 

Set Schema file or URL to https://github.com/alexiswl/cwl-ts-auto/blob/enhancement/generate-json-schema/cwl_schema.json

Set Schema Version to 'JSON Schema version 7'

Click Apply

image

## VSCode

1. Install the RedHat YAML Extension (this will enable the `yaml.schemas` key in the settings.json

image


2. Open the settings json file and paste in the following keys

Screenshot 2024-02-13 at 3 02 49 pm


The settings.json can be accessed via the command pallete.  

We also need to update the files.associations so that .cwl files are seen as YAML files by the compiler.

```json
{
    "yaml.schemas": {
        "https://raw.githubusercontent.com/alexiswl/cwl-ts-auto/enhancement/generate-json-schema/cwl_schema.json": "*.cwl"
    },
    "files.associations": {
        "*.cwl": "yaml"
    }
}
```

image



## Uses

### Highlight invalid keys

image

image


image

image

image

image


### Show documentation

image

image

image



### Provide Valid Suggestions

image

image

image

image

**Comment by alexiswl** [comment]
Ah makes sense, this is a cross-fork write out error. 

Can we merge this into a branch on the origin repo first and then retry the GH action test? I don't think I have permission to make a branch on this repository

**Comment by mr-c** [comment]
@alexiswl I sent you an invitation to join https://github.com/orgs/common-workflow-lab/teams/people/members which gives you permission to make branches in this and many other CWL repos. Feel free to re-open this PR from a local branch and thanks again for your contributions!

File: cwl_schema.json [hunk 1]
```diff
@@ -0,0 +1,6270 @@
+{
+    "$ref": "#/definitions/CWLFile",
+    "$schema": "http://json-schema.org/draft-07/schema#",
```
Comment (mr-c) [inline cwl_schema.json:3]: Are there any advantages to producing a schema using the newer drafts of the JSON schema spec?

Comment (alexiswl) [reply cwl_schema.json:3]: Potentially, JetBrains only seems to support version 7 though from the drop down menus, but I don't know if they actually support more recent versions (2019 and 2020). From the JSON Schema Store, GitHub Actions schema is also on version 7 still

**Comment by alexiswl** [comment]
Updated base branch to equivalent branch name on the common-workflow-lab/cwl-ts-auto repo.  

Merging and will then create the PR to main from the common-workflow-lab branch instead.

**Pull request merged**